// webpacks/src/dev-server.js
'use strict'

module.exports = {
  setup: function (server, config, options) {
    options = options || {}
    config.entry.app = ['webpack-hot-middleware/client', config.entry.app]
    config.plugins.push(
      new webpack.HotModuleReplacementPlugin(),
      new webpack.NoErrorsPlugin()
    )

    var compiler = require('webpack')(config)
    var devMiddleware = require('webpack-dev-middleware')(compiler, {
      publicPath: config.output.publicPath,
      stats: {
        colors: true,
        modules: false,
        children: false,
        chunks: false,
        chunkModules: false
      }
    })
    server.use(devMiddleware)
    server.use(require('webpack-hot-middleware')(compiler))

    if(options.callback){

      var assetReader = {
        read: function(assetPath){
          return devMiddleware.fileSystem.readFileSync(path.join(config.output.path, assetPath), 'utf8')
        }
      }
      options.callback({
        devMiddleware: devMiddleware,
        assetReader: assetReader
      })
    }
  }
}
